src/Bicep.LangServer/Handlers/BicepSignatureHelpHandler.cs

```diff
@@ -0,0 +1,271 @@
+            TryReuseActiveSignature(request.Context, signatureHelp);
+
+            return Task.FromResult<SignatureHelp?>(signatureHelp);
+        }
+
+        private static SyntaxBase? GetActiveFunctionCall(ProgramSyntax syntax, int offset, out ImmutableArray<FunctionArgumentSyntax> arguments)
+        {
+            // if the cursor is placed after the closing paren of a function, it needs to count as outside of that function call
+            // for purposes of signature help (otherwise we'll show the wrong function when function calls are nested)
+            var matchingNodes = SyntaxMatcher.FindNodesMatchingOffsetExclusive(syntax, offset);
+
+            var index = matchingNodes
+                .FindLastIndex(
+                    matchingNodes.Count - 1,
+                    current => current is FunctionCallSyntax functionCall && TextSpan.BetweenExclusive(functionCall.OpenParen.Span, functionCall.CloseParen).ContainsInclusive(offset) ||
+                               current is InstanceFunctionCallSyntax instanceFunctionCall && TextSpan.BetweenExclusive(instanceFunctionCall.OpenParen.Span, instanceFunctionCall.CloseParen.Span).ContainsInclusive(offset));
+
+            var activeFunctionCall = index < 0 ? null : matchingNodes[index];
+            switch (activeFunctionCall)
+            {
+                case FunctionCallSyntax functionCall:
+                    arguments = functionCall.Arguments;
+                    return functionCall;
+
+                case InstanceFunctionCallSyntax instanceFunctionCall:
+                    arguments = instanceFunctionCall.Arguments;
+                    return instanceFunctionCall;
+
+                default:
+                    arguments = ImmutableArray<FunctionArgumentSyntax>.Empty;
+                    return null;
+            }
+        }
+
+        private static void TryReuseActiveSignature(SignatureHelpContext? context, SignatureHelp signatureHelp)
+        {
+            if (context?.ActiveSignatureHelp == null ||
+                string.Equals(context.TriggerCharacter, FunctionArgumentStart, StringComparison.Ordinal) ||
+                string.Equals(context.TriggerCharacter, FunctionArgumentEnd, StringComparison.Ordinal))
+            {
+                // we don't have a previous active signature or the user typed ( or ), which would indicate a new "session"
+                return;
+            }
+
+            if (CheckIfSignatureHelpSimilar(context.ActiveSignatureHelp, signatureHelp))
+            {
+                // the signature help is for the same function so we can reuse the active signature index
+                // this prevents resetting of the active signature when multiple overloads are ambiguous and the user selected a specific one manually
+                signatureHelp.ActiveSignature = context.ActiveSignatureHelp.ActiveSignature;
+            }
+        }
+
+        private static bool CheckIfSignatureHelpSimilar(SignatureHelp active, SignatureHelp @new)
```

Comment:
>CheckIfSignatureHelpSimilar [](start = 28, length = 27)

The implementation here is super hacky. Ideally LSP would give us ability to put in arbitrary metadata like we can on `CompletionItem`, but I think we'll need to submit a proposal for that.